JavaScriptCodeConventions.md

```diff
@@ -0,0 +1,70 @@
+# JavaScript Code Conventions
+> NOTE: These conventions also apply to super-sets of JavaScript which are used within Ultimaker such as TypeScript.
+
+## Introduction
+These JavaScript style guidelines are intended to be used by both the R&D and marketing team as part of a broader effort to standardize JavaScript style across the company and as an early facilitator for the fact that Ultimaker continues to offer more web-based products such as the Cura Cloud services and Cura Connect.
+
+## AirBnB as Baseline Style
+To reduce the need to build comprehensive style guidelines ourselves, we have decided to adopt the well-known and widely used **AirBnB style**. The full documentation for AirBnB style [can be found on GitHub](https://github.com/airbnb/javascript). For the most part, this style is already very similar to what is used by front-end teams at Ultimaker with a few differences, which are outlined below. It was also chosen because of it was designed for usage with React.js, which is used by the R&D department. It is applicable to Vue.js as well, however, and can be used by the marketing department.
```

Comment:
AirBnB style conflicts with our [global code conventions](https://github.com/Ultimaker/Meta/blob/master/code_conventions.md) in certain areas. Please clarify.

Reply:
Each language file serves as an addendum to the global code conventions, right? So AirBnB is specific to JavaScript code and covers more than the global code conventions do. The only cases I know of so far where the AirBnB style conflicts with existing JS projects at Ultimaker are outlined below. But Cloud, Connect, and Marketing all already pretty much follow AirBnB so it's not so complicated to say "just follow it 100% with 1 or 2 exceptions" (like 4 spaces instead of 2).

Reply:
So far, we have the Python code style conventions which [specify that our global code conventions override PEP8](https://github.com/Ultimaker/Meta/blob/master/PythonCodeConventions.md#python-code-conventions), and the C++ code style conventions which don't have any conflicts with the global code conventions.

As far as I can see, AirBnB also conflicts with us in these aspects:
* Brackets on new lines.
* Strings using double quotes.
* Leading underscores in names.
* Using lower_case_underscores.
* Every class in its own file.

Reply:
> - Leading underscores in names.

These are addressed within the proposal.

> - Brackets on new lines.
> - Strings using double quotes.
> - Using lower_case_underscore naming for local variables and parameters.

It's assumed the language specific guidelines supersede the general guidelines, no? In many cases we do this already. You know my thoughts on how the first two are applied 50/50 in, say, our QML, and the 3rd is already completely unused at all in our existing JS or QML code.

> - Every class in its own file.

No conflict here. AirBnB states that you should use a single default export for each file (i.e., impossible to use two classes from a single file.)

Reply:
The language specific ones can override the general one, but this should be done sparingly. 

The fact that it's not done does not mean that we should not define it (or adhere to our own standards).

Reply:
I agree with Ian on the other points, that this should override those three points from the overall code standard. We want to keep the exceptions to a minimum and we already have two (indent with four spaces instead of two, and allowing variables to start with an underscore). We want to stay as close as possible to what the JS community generally does to avoid difficultly for new hires getting aligned. I don't think I've ever seen code examples in JS that use brackets on new lines after statements or underscores in between words in variables, for example none of the standards we looked at recommend this, so it doesn't make sense to impose that on people coming from a JS background. @Ghostkeeper can we agree that there can be some differences between programming languages?

Reply:
The argument that it would impede new hires is weak at best; All of our python code is also not adhering to the most commonly used style. The same can also be said for C++.
I personally feel that an overall consistency is much more important, especially when switching between projects (which, although it shouldn't happen, does happen). This is the main reason that we originally went for a global codestyle. I think it's a bit weird to suddenly start diverting from it (eg; It's way more confusing to have a completely different codestyle on one of the languages compared to the global one because that's even more rules to learn.)

Reply:
> I personally feel that an overall consistency is much more important, especially when switching between projects.

That's what we're trying to address.

> I think it's a bit weird to suddenly start diverting from it (eg; It's way more confusing to have a completely different codestyle on one of the languages compared to the global one because that's even more rules to learn.)

Sorry... I'm just going to say it because I'm tired of thinking it and not saying it: "That's stupid." Different languages require different grammar and syntax and trying to apply a global style which is both ill suited and doesn't appear to have taken a language into consideration to a given language is dumb. In the case of JavaScript we follow a weird blend of Python and C++ conventions, neither of which really fit, nor make sense. They're also not well documented and there's no set-up right now to automatically enforce the style with ESLint. This is a road block to working between teams or with outside parties, which is what spurred this discussion in the first place.

*The global style does not make things clear and simple for people. It makes things muddled. Adopting a well-known and impressively well documented standard which we pretty much already follow, does.*

Reply:
Also, I know of only 3 points where AirBnB conflicts (leading underscores, indent size, quotes) we defer to Cura global on 2 and to AirBnB on 1. Sounds like "addendums" to me.

Reply:
They are documented, we recently changed this. There is also not a single reason why those codestyles would not work for javascript. Prepending things with a _ is not required (but also not in the global codestyle). The global codestyle is about structure and wether we use camal casing. Did you even read what is in there? Based off your comments, I don't have the impression that you did. 

Adopting it only in a single place will, at best, make your project better. The main problem is that it does so at the expense of other printers. This simply feels like some engineers felt left out with the original design of the codestyle and see this as a chance to change it in a way they see fit. 

Your main argument is that the style has not been enforced. You're right. But that is outside the scope of the rules. We should enforce them so that all code has a style that is alike. This change will not suddenly make the mix any better, it will make it **worse**. Instead of it being poorly policed rules in the case of Cura, it will be muddled rules (effectively moving the problem one step up). Changing this will not just affect what you are doing, it affects other teams as well.

Reply:
> Adopting it only in a single place will, at best, make your project better.

We want to adopt this on all projects which use JavaScript.

> This simply feels like some engineers felt left out with the original design of the codestyle and see this as a chance to change it in a way they see fit.

OK, let me put it this way: Everyone I've spoken to who works in this language thinks the global rules are silly and irritating to use on a daily basis. They have been widely disregarded because of this. And no, I'm not talking about Ian doesn't like how it's done in QML, I'm talking about the entirety of the react-web-components repo, the entirety of the print-cluster repo, and the entirety of the marketing department. What is here *actually has the consensus of those who work on JavaScript within Ultimaker.* Why does accommodating people bother you so much? Do I need to go around and collect signatures?

> Your main argument is that the style has not been enforced.

Nope. But I've said my (many) arguments many times and I'll save it for the architecture guild before I go crazy from repeating myself.

Reply:
Since no-one from Cura, who also work in javascript, have been asked (or agree), I think it's a bit fast to talk about consensus. There is no need to collect signatures, I don't doubt that this is how most javascript is done. I already gave that argument in a previous post (eg; our python code is also not following how most python developers do it, so i feel that's a bit of a non argument). This is the exact same discussion we had when we first started codestyles. We opted back then to take a global style in favor of completely language specific ones.

So this has nothing to do with me wanting to mess with you, it is me being tired of having to add exceptions to the codestyle every year or so because new people started working at Ultimaker.